app/code/Magento/Catalog/Block/Product/View.php

```diff
@@ -173,7 +181,12 @@ public function getProduct()
             $product = $this->productRepository->getById($this->getProductId());
             $this->_coreRegistry->register('product', $product);
         }
-        return $this->_coreRegistry->registry('product');
+        $product = $this->_coreRegistry->registry('product');
+        if (! $product) {
```

Comment:
Please remove space, it is not strictly forbidden yet but inconsistent with other Magento code.

Please avoid unnecessary access to registry.

Please cover this new logic with unit test if it's not too hard.